torch/utils/data/datapipes/_typing.py

```diff
@@ -371,9 +371,12 @@ def wrap_setstate(*args, **kwargs):
                 Set `_restored` to True during `__setstate__`, such that the next `reset()` call during
                 iterator creation will not actually reset the state of the DataPipe.
                 """
-                datapipe = args[0]
-                datapipe._restored = True
-                return setstate_func(*args, **kwargs)
+                try:
+                    return setstate_func(*args, **kwargs)
+                finally:
+                    # This needs happen after because the state being passed in may have `datapipe._restored = False`.
+                    datapipe = args[0]
+                    datapipe._restored = True
```

Comment:
Rewriting the `__setstate__` hook logic because the previous version doesn't work with serialization. In particular, `datapipe._restored` is set to `False` because `setstate_func` overwrites it.